### PR TITLE
PHPword: insert chart in template

### PR DESCRIPTION
**Pull request by dbarzin**
### Description

Add a function in the Word templateProcessor 

Fixes # 1058

### Checklist:

- [x] I have run `composer run-script check --timeout=0` and no errors were reported
- [x] The new code is covered by unit tests (check build/coverage for coverage report)
- [x] I have updated the documentation to describe the changes

### CODE REVIEW

**Comment by snazy2000** [comment]
@dbarzin just tried this from your repo but even the sample just returns a corrupt docx file?

**Comment by dbarzin** [comment]
Hi Stephen,

Have you tried this sample code ?

samples/Sample_41_TemplateSetChart.php

I use Ubuntu with OpenOffice.
I will test the result on Windows with word tomorrow.

Thanks
Didier

**Comment by dbarzin** [comment]
Check made from my github and it is working fine. Here is what I have done :

git clone https://github.com/dbarzin/PHPWord
cd PHPWord 
composer install
cd samples
php ./Sample_41_TemplateSetChart.php

The generated file is in results/Sample_41_TemplateSetChart.docx

**Comment by dbarzin** [comment]
#1865 Colors not applied correcty in stacked_column charts

**Comment by dbarzin** [comment]
fixes #1865 Colors not applied correcty in stacked_column charts 

<?php
include_once 'Sample_Header.php';

use PhpOffice\PhpWord\Shared\Converter;

// New Word document
echo date('H:i:s'), ' Create new PhpWord object', EOL;
$phpWord = new \PhpOffice\PhpWord\PhpWord();

// 2D charts
$section = $phpWord->addSection();
$section->addTitle('2D charts', 1);
$section = $phpWord->addSection(array('colsNum' => 2, 'breakType' => 'continuous'));

$colors = array(
	'FF0000','FF0000','FF0000','FF0000','FF0000',
	'00FF00', '00FF00', '00FF00', '00FF00', '00FF00', 
	'0000FF', '0000FF', '0000FF', '0000FF', '0000FF'
	);
$categories = array('A', 'B', 'C', 'D', 'E');
$series1 = array(1, 3, 2, 5, 4);
$series2 = array(3, 1, 7, 2, 6);
$series3 = array(8, 3, 2, 5, 4);

$chart = $section->addChart('stacked_column', $categories, $series1);
$chart->addSeries($categories, $series2);
$chart->addSeries($categories, $series3);

$chart->getStyle()->setWidth(Converter::inchToEmu(5))->setHeight(Converter::inchToEmu(3));
$chart->getStyle()->setColors($colors);

// Save file
echo write($phpWord, basename(__FILE__, '.php'), $writers);
if (!CLI) {
    include_once 'Sample_Footer.php';
}
?>

**Comment by snazy2000** [comment]
I was doing it directly from the browser and when it generated the document it hasn't put anything inside the content_types which is why it was then corrupt

I had to change it to the following to get it to work

https://github.com/snazy2000/PHPWord/commit/78a3eb7933ef4d4bfacd580d5bd3e26d4b5d893b

Also correct me if I'm wrong but your most recent commit about colours looks like you've removed all the ability to write charts? lol Looks like stamped over Part\chart.php with Element\chart.php :)

**Comment by dbarzin** [comment]
The invalid commit has been reverted.

Have you tried the sample 41(Sample_41_TemplateSetChart.php)  ? For an easy access to those samples launch php -S localhost:8000 in the samples directory then browse to http://localhost:8000 to view the samples.

**Comment by snazy2000** [comment]
Yes I have and it doesn't work in Windows Word, haven't tried Openoffice.

**Comment by dbarzin** [comment]
I confirm that the file generated could not be open with Word. I try to
find out why this week.

**Comment by dbarzin** [comment]
I have pushed a fix

File: src/PhpWord/Writer/Word2007/Part/Chart.php [hunk 1]
```diff
@@ -265,23 +266,21 @@ private function writeSeries(XMLWriter $xmlWriter, $scatter = false)
...
-                    // This is a workaround to make each series in a stack chart use a different color
-                    if ($this->options['type'] == 'bar' && stristr($this->options['grouping'], 'stacked')) {
-                        array_shift($colors);
-                    }
-                    $colorIndex = 0;
-                    foreach ($colors as $color) {
+                // check that there are colors
+                if (is_array($colors) && count($colors)>0) {
+                    // assign a color to each value
+                    $valueIndex=0;
```
Comment (troosan) [inline src/PhpWord/Writer/Word2007/Part/Chart.php:272]: is this workaround for stacked charts not needed anymore?

Comment (troosan) [reply src/PhpWord/Writer/Word2007/Part/Chart.php:272]: how did you fix this? without the shift for stacked bars the unit test testChartElements in ChartTest fails.

Comment (dbarzin) [reply src/PhpWord/Writer/Word2007/Part/Chart.php:272]: I do not alter the array but get the next color using modular computation :
$xmlWriter->writeElementBlock('a:srgbClr', 'val', $colors[$colorIndex++ % count($colors)]);

**Comment by troosan** [comment]
will be merged with https://github.com/PHPOffice/PHPWord/pull/2012

**Pull request closed without merging**